### PR TITLE
Fix crash when analyzing shaders in devkit

### PR DESCRIPTION
**Pull request by Filoppi**
the functions were returning a char pointer within an object that might have already been destroyed

### CODE REVIEW

**Review by clshortfuse — CHANGES_REQUESTED** [review]

File: src/utils/shaderCompiler.hpp [hunk 1]
```diff
@@ -65,29 +63,32 @@ namespace ShaderCompilerUtil {
...
 
-    return (char*)ppDisassembly->GetBufferPointer();
+    return disassembly;
   }
 
-  char* disassembleShader(void* code, size_t size) {
-    char* result = disassembleShaderFXC(code, size);
+  std::string disassembleShader(void* code, size_t size) {
+    CComPtr<ID3DBlob> result = disassembleShaderFXC(code, size);
     if (result == nullptr) {
```
[the remark this answers is not in the document]
Comment (Filoppi) [reply src/utils/shaderCompiler.hpp:84]: yes

File: src/devkit/addon.cpp [hunk 2]
```diff
@@ -855,7 +855,7 @@ static void on_init_pipeline(
     reshade::log_message(reshade::log_level::info, s.str().c_str());
   }
   if (!foundUsefulShader) {
-    free(newSubobjects);
+    delete newSubobjects;
```
[the remark this answers is not in the document]
Comment (Filoppi) [reply src/devkit/addon.cpp:858]: it's not. I just replaced all free with delete. AFAIK it's always better?

File: src/devkit/addon.cpp [hunk 3]
```diff
@@ -1888,12 +1888,11 @@ static void on_register_overlay(reshade::api::effect_runtime* runtime) {
               auto hash = traceHashes.at(selectedIndex);
               auto cache = shaderCache.find(hash)->second;
               if (cache->disasm.length() == 0) {
-                char* disasmCode = ShaderCompilerUtil::disassembleShader(cache->data, cache->size);
-                if (disasmCode == nullptr) {
```
[the remark this answers is not in the document]
Comment (Filoppi) [reply src/devkit/addon.cpp:1892]: the source of the crash is that the `ID3DBlob` could end up being de-allocated at any time, and we were trying to "free" it ourselves, which isn't allowed. `ID3DBlob` lifespan needs to be handled with `CComPtr<ID3DBlob>`.

[the remark this answers is not in the document]
Comment (clshortfuse) [reply src/devkit/addon.cpp:1892]: That's not true. The Blob gets copied and released.

[the remark this answers is not in the document]
Comment (Filoppi) [reply src/devkit/addon.cpp:1892]: It wasn't ever released.

[the remark this answers is not in the document]
Comment (clshortfuse) [reply src/devkit/addon.cpp:1892]: You're, right I was looking at `compileShaderFromFileFXC` which does copy the output

**Pull request closed without merging**